File: AoC2017/Day02.fs
namespace AoC2017

open Common
open System.Collections.Generic
open System.Linq 

type Day02 = 
    inherit Solution

    new (file:string) = { inherit Solution(file, "\n") }
    new (input : List<string>) = { inherit Solution(input) }

    override this.SolvePart1() = 
        this.Input
            .Select(fun line -> line.Split("\t").ToList().Select(fun e -> e |> int32))
            .Aggregate(0, fun agg row -> agg + (row.Max() - row.Min())).ToString()
           
    override this.SolvePart2() = 
        this.Input
            .Select(fun line -> line.Split("\t").ToList().Select(fun e -> e |> int32).ToArray())
            .Aggregate(0, fun agg row -> agg + this.rowSum(row)).ToString()

    member this.rowSum(row: int32[] ): int32 = 
        let mutable sum = 0
        for i in 0..row.Length-2 do 
            for j in i+1..row.Length-1 do 
                let iSmallest = row.[i] <= row.[j]
                let m = if iSmallest then row.[j] % row.[i] else row.[i] % row.[j]              
                sum <- if m = 0 && iSmallest then (row.[j] / row.[i] ) elif m = 0 then (row.[i] / row.[j] ) else sum
        sum
            

File: AoC2017/Day02.cs
using Common;
using Common.Extensions;

namespace AoC2017;


public class Day02 : Solution
{
	private List<List<float>> SpreadSheet;

	public Day02(string file) : base(file) => SpreadSheet = Input.Select(l => l.Split('\t').Select(float.Parse).ToList( )).ToList( );


	public override async Task<string> SolvePart1() => SpreadSheet
		.Select(l => l.Max( ) - l.Min( )).Sum( ).ToString( );


	public override async Task<string> SolvePart2() => SpreadSheet
		.Select(r => r.OrderDescending( ).WithIndex( ).ToList( ))
		.Aggregate(0f, (sumOverall, row) => sumOverall + row.SkipLast(1)
			.Aggregate(0f, (sum, entry) => sum + Enumerable.Range(entry.idx + 1, row.Count - entry.idx - 1)
				.Aggregate(0f, (s, idx) => float.IsInteger(entry.Value / row[idx].Value) ? s + entry.Value / row[idx].Value : s)
		)).ToString( );

}
